### PR TITLE
bug: Add support for access to eks cluster with private endpoint

### LINKED ISSUE #2465: Private EKS cluster SG forbids by default API access from outside of cluster
/kind bug

**What steps did you take and what happened:**
If you create EKS cluster with:
```yaml
  endpointAccess:
    public: false
    private: true
```
Than default security group setup is done in this way:
* Controll plane primary SG:
  * Ingress
    * All Traffic from self (reference to same SG id)
  * Egress
    * All Traffic to 0.0.0.0/0 
* Controll plane secondary SG:
  * Ingress
    * Port 22 TCP from SG of bastion 
* Nodegroup:
  * uses same  SG as Controll plane primary SG

**What did you expect to happen:**
This works for public cluster to an extent but is still weird setup. If you have private cluster it basically means noone except cluster itself can talk to API. Coupled with another bug which is `networkSpec: {  securityGroupOverrides: }` is not working for EKS which will be raised as different bug makes really private cluster completely unusable.

**Anything else you would like to add:**

**Environment:**

- Cluster-api-provider-aws version: 
```sh
$ kubectl get deploy -n capa-system capa-controller-manager -o yaml | grep cluster-api-aws-controller
        image: gcr.io/k8s-staging-cluster-api-aws/cluster-api-aws-controller:v0.6.6
```
- Kubernetes version: (use `kubectl version`): 
```sh
$ kubectl cluster-info
Kubernetes control plane is running at https://127.0.0.1:39569
KubeDNS is running at https://127.0.0.1:39569/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy
```
- OS (e.g. from `/etc/os-release`): 
```sh
$ cat /etc/os-release
NAME="Amazon Linux"
VERSION="2"
ID="amzn"
ID_LIKE="centos rhel fedora"
VERSION_ID="2"
PRETTY_NAME="Amazon Linux 2"
ANSI_COLOR="0;33"
CPE_NAME="cpe:2.3:o:amazon:amazon_linux:2"
HOME_URL="https://amazonlinux.com/"
```

### PR DESCRIPTION
**Pull request by paulcarlton-ww**
Add support for allowing access to private API endpoint from the
bastion host and also from specified security groups.

Fixes issue #2465

Signed-off-by: paulcarlton-ww <paul.carlton@weave.works>

**What type of PR is this?**

**What this PR does / why we need it**:

Without this feature the management cluster cannot access the API of a tenant cluster with only a private endpoint. Tenant cluster API access is required to complete provisioning.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2465

**Special notes for your reviewer**:
Untested, need to revisit this and add unit tests but need to prioritize other work right now 

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Adds new fields to the `AWSManagedControlPlane` `endpointAccess` element to allow access to a private API endpoint 

endpointAccess:
    public: false
    private: true
    bastionAccess: true
    additionalSecurityGroups:
    - "..."
```
Note that the use of `additionalSecurityGroups` to add security groups that allow management cluster access to the tenant's private API may require the creation of a VPC peering connection, see #2484

### CODE REVIEW

**Comment by randomvariable** [comment]
Need to run 

```
make generate
```

You also will need to add v1alpha3 conversion using the annotation marshalling/unmarshalling.

**Comment by richardcase** [comment]
@paulcarlton-ww - ping me on slack if you want a hand with the conversion part

File: controlplane/eks/api/v1alpha4/awsmanagedcontrolplane_types.go [hunk 1]
```diff
@@ -173,6 +173,12 @@ type EndpointAccess struct {
 	// Private points VPC-internal control plane access to the private endpoint
 	// +optional
 	Private *bool `json:"private,omitempty"`
+	// BastionAccess controls whether API access is available from the bastion host.
+	// +optional
+	BastionAccess *bool `json:"bastionAccess,omitempty"`
+	// AdditionalSecurityGroups specifies additional securirty groups that should be granted API access.
+	// +optional
+	AdditionalSecurityGroups []*string `json:"additionalSecurityGroups,omitempty"`
```
Comment (richardcase) [inline controlplane/eks/api/v1alpha4/awsmanagedcontrolplane_types.go:181]: We are moving away from a slice of pointers to structs due to an issue with conversion-gen (see #2495). Could you change to `[]string` and then check for the length of this elsewhere.

**Review by richardcase — CHANGES_REQUESTED** [review]

**Comment by paulcarlton-ww** [comment]
> @paulcarlton-ww - ping me on slack if you want a hand with the conversion part

Will do, thanks @richardcase
Busy with other stuff right now will get back to this as soon as able

File: pkg/cloud/services/eks/cluster.go [hunk 2]
```diff
@@ -264,6 +264,14 @@ func makeVpcConfig(subnets infrav1.Subnets, endpointAccess controlplanev1.Endpoi
 	if ok {
 		vpcConfig.SecurityGroupIds = append(vpcConfig.SecurityGroupIds, &sg.ID)
 	}
+
+	if *(endpointAccess.BastionAccess) {
+		sgID := securityGroups[infrav1.SecurityGroupBastion].ID
```
Comment (shivi28) [inline pkg/cloud/services/eks/cluster.go:269]: Handle nil pointer exception
```
if  securityGroup, okay := securityGroups[infrav1.SecurityGroupBastion]; okay{
  endpointAccess.AdditionalSecurityGroups = append(endpointAccess.AdditionalSecurityGroups, &securityGroup.ID)
}
```

**Comment by paulcarlton-ww** [comment]
Superseded by #2749

**Pull request closed without merging**